pinot-query-planner/src/main/java/org/apache/pinot/query/type/TypeFactory.java

```diff
@@ -0,0 +1,74 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.type;
+
+import java.util.Map;
+import org.apache.calcite.jdbc.JavaTypeFactoryImpl;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeSystem;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+
+
+/**
+ * Extends Java-base TypeFactory from Calcite.
+ */
+public class TypeFactory extends JavaTypeFactoryImpl {
+  private final RelDataTypeSystem _typeSystem;
+
+  public TypeFactory(RelDataTypeSystem typeSystem) {
+    _typeSystem = typeSystem;
+  }
+
+  public RelDataType createRelDataTypeFromSchema(Schema schema) {
+    Builder builder = new Builder(this);
+    for (Map.Entry<String, FieldSpec> e : schema.getFieldSpecMap().entrySet()) {
+      builder.add(e.getKey(), toRelDataType(e.getValue()));
+    }
+    return builder.build();
+  }
+
+  private RelDataType toRelDataType(FieldSpec fieldSpec) {
+    switch (fieldSpec.getDataType()) {
+      case INT:
+        return this.createSqlType(SqlTypeName.INTEGER);
+      case LONG:
+        return this.createSqlType(SqlTypeName.BIGINT);
+      case FLOAT:
+        return this.createSqlType(SqlTypeName.FLOAT);
+      case DOUBLE:
+        return this.createSqlType(SqlTypeName.DOUBLE);
+      case BOOLEAN:
+        return this.createSqlType(SqlTypeName.BOOLEAN);
+      case TIMESTAMP:
+        return this.createSqlType(SqlTypeName.TIMESTAMP);
+      case STRING:
+        return this.createSqlType(SqlTypeName.VARCHAR);
+      case BYTES:
+        return this.createSqlType(SqlTypeName.VARBINARY);
+      case JSON:
```

Comment:
JSON is a recognized type in Pinot so we should not throw Unsup for that ?

Reply:
JSON will also be implemented as Struct type i suppose.

Reply:
TODO is fine for now I guess. We must handle JSON since it's a first class type in Pinot now